### PR TITLE
FIX base theme initialization and theme bootup

### PR DESCRIPTION
**Pull request by ndelangen**
Issue: #5824 #5780 

creating a theme that derived off of `light` or `dark` default themes was impractical, and wasn't working quite as users were expecting it to behave.

## What I did
ADD ability for users to specify baseTheme, and the theme createFn will merge in the appropriate themeVars.

### CODE REVIEW

File: lib/theming/src/themes/light.ts [hunk 1]
```diff
@@ -1,36 +1,10 @@
 import { create } from '../create';
-import { color, typography, background } from '../base';
+
+import { themeVars } from './dark-vars';
```
Comment (shilman) [inline lib/theming/src/themes/light.ts]: @ndelangen should this be from `light-vars`?

**Review by shilman — CHANGES_REQUESTED** [review]
@ndelangen please test the hell out of this?

File: lib/theming/src/themes/dark.ts [hunk 2]
```diff
@@ -1,36 +1,9 @@
...
-  barTextColor: '#999999',
-  barSelectedColor: color.secondary,
-  barBg: color.darkest,
-
-  // Form colors
-  inputBg: '#3f3f3f',
-  inputBorder: 'rgba(0,0,0,.3)',
-  inputTextColor: color.lightest,
-  inputBorderRadius: 4,
+  ...themeVars,
```
Comment (tmeasday) [inline lib/theming/src/themes/dark.ts]: Isn't this superfluous?

File: lib/theming/src/create.ts [hunk 3]
```diff
@@ -110,76 +115,85 @@ const darkSyntaxColors = {
...
+
+export const create = (vars: ThemeVar, rest?: Rest): Theme => {
+  const inherit: ThemeVar = { ...vars, ...(base[vars.base] || base.light), ...base.light };
+  const p = pick.bind(null, inherit);
+
+  return {
+    base: vars.base,
+    color: createColors(inherit),
+    background: {
+      app: p('appBg'),
```
Comment (tmeasday) [inline lib/theming/src/create.ts]: I don't really understand why this isn't just `inherit.appBg`?

File: lib/theming/src/create.ts [hunk 4]
```diff
@@ -110,76 +115,85 @@ const darkSyntaxColors = {
...
-  typography: {
-    fonts: {
-      base: vars.fontBase || typography.fonts.base,
-      mono: vars.fontCode || typography.fonts.mono,
+function pick<B, K extends keyof B>(b: B, key: K): B[K] {
+  return b[key];
+}
+
+export const create = (vars: ThemeVar, rest?: Rest): Theme => {
+  const inherit: ThemeVar = { ...vars, ...(base[vars.base] || base.light), ...base.light };
```
Comment (tmeasday) [inline lib/theming/src/create.ts]: Should this just be `themeVars` or something

**Review by tmeasday — CHANGES_REQUESTED** [review]
I feel like this code would benefit from some tests but I appreciate this is a fairly late change

File: lib/theming/src/create.ts (hunk 4 above)
Comment (tmeasday) [reply lib/theming/src/create.ts]: I mean this is just a naming question, not a big deal ;)

File: lib/theming/src/themes/dark.ts (hunk 2 above)
Comment (tmeasday) [reply lib/theming/src/themes/dark.ts]: As in, wouldn't `create({ base: 'dark' })` have the exact same effect, by definition.

**Comment by shilman** [comment]
When I add this line to `offcial-storybook/config.js` the app hangs:

```
    theme: create({ base: 'dark', colorPrimary: '#FF4785', colorSecondary: '#1EA7FD' }),
```

File: lib/theming/src/themes/light.ts (hunk 1 above)
Comment (ndelangen) [reply lib/theming/src/themes/light.ts]: indeed

File: lib/theming/src/create.ts (hunk 3 above)
Comment (ndelangen) [reply lib/theming/src/create.ts]: hmm, yeah

**Comment by ndelangen** [comment]
I'm writing tests @shilman

**Comment by shilman** [comment]
@ndelangen A couple comments, both outside the scope of this PR:
- After testing out a buggy version, I had to do `localStorage.clear()` to get the theme to load. Seems like a bug, but maybe it's a corner case
- Seems like the theme serialization should be profiled.. pretty slow to load on my laptop

**Review by shilman — CHANGES_REQUESTED** [review]
@ndelangen Maybe a stupid question:

```
    theme: create({ base: 'dark', colorPrimary: 'red', colorSecondary: 'green' }),
```

I did this in `official-storybook`. I see green in the resulting UI, but I don't see any red. Doesn't seem right to me?

**Comment by ndelangen** [comment]
@shilman 

I'd like to remove the need to do theme serialization asap.

theme.color.primary seems to be very rarely used, @domyen came up with the naming and usage of colors.

**Comment by domyen** [comment]
@ndelangen yea `theme.color.primary` is not used at all, happy for you to axe it! (the var was mirrored from storybook/frontpage, but we don't need it here)

**Comment by shilman** [comment]
@domyen @ndelangen seems like an unfortunate choice of names then... I thought the point of this minimal theme var abstraction was to help us preserve backwards compatibility?

**Comment by shilman** [comment]
We should be making minimal changes at this point though, so I don't really want to change it. Though it makes me cringe a bit.

**Comment by domyen** [comment]
I’m talking about what’s exposed in the create() function “colorPrimary”.

We still want to have a theme.color.primary as a part of the larger theme
colors (it’s set to storybook pink).

**Comment by ndelangen** [comment]
@shilman @tmeasday 

I did a refactor to ensure the least amount of data needed is transferred over the channel.

I ended up not 'creating' the full theme and then trying to transmit, instead it's much more efficient to transfer the `ThemeVars` across and construct the actual `Theme` in the manager.

This means no more hack to make sure the functions make it across, since is a plain flat object.

I know it's not what you were hoping for: it's a bigger change I wish was necessary.

**Comment by ndelangen** [comment]
I wrote in such a way, users won't need to change anything.
As long as they are using `create()`..

**Comment by ndelangen** [comment]
I also noticed that setOptions is called a ton, and it's was re-creating the theme on every navigation.

It'd be nice to fix this a bit better, but for now I improved it by `deepEqual(theme, updatedTheme)`.

**Pull request merged**